### ISSUE TITLE
Support for ASIO

### ISSUE THREAD

**Issue by siraaris** [HEnquist/camilladsp#346]
Devices on Windows with more than 8 ch are presented as multiple WDM devices with 8 ch each. 

In contrast the ASIO driver presents the whole channel count. 

For these devices, CamillaDSP can't be used on Windows. 

Would it be possible to add a build option ASIO backend?

I understand that ASIO and GPLv3 licensing does not support distribution of a binary, but a self-build option would work for people who need it. 

I had a brief search for Rust ASIO libraries, some seem reasonable.

**Comment by HEnquist** [comment]
> Devices on Windows with more than 8 ch are presented as multiple WDM devices with 8 ch each.
> 

This is not determined by WDM, there is no 8 channel limit. The manufacturer has simply chosen to present the device like this. This is quite common (and also quite annoying). The Motu M4 for example is presented as two stereo devices.
 

> Would it be possible to add a build option ASIO backend?
> 
> I understand that ASIO and GPLv3 licensing does not support distribution of a binary, but a self-build option would work for people who need it.
> 
> I had a brief search for Rust ASIO libraries, some seem reasonable.

ASIO has a number of problems in addition to the license. See here: https://www.diyaudio.com/community/threads/camilladsp-cross-platform-iir-and-fir-engine-for-crossovers-room-correction-etc.349818/post-6813205
There are supposedly ways to work around these problems but implementing a new backend is already a lot of work without having to deal with such stuff. The only asio binding I'm aware of is the low level `asio-sys` that is used in the CPAL library. I don't really have the time to implement and maintain an asio backend.  

Then one reasonable path I see is to use the asio support of CPAL. This would be quite straightforward since CPAL is already used by the jack backend. Unfortunately this worked very poorly last time I tried it. I'm following the development there, but asio doesn't seem to be a priority so not much happens.

**Comment by siraaris** [comment]
My device is a RME Digiface AVB, and inherently has a large channel count, and I use it to access a M32-DA Pro DAC which has 32 ch analog output. 

I can't use CamillaDSP for this on Linux yet but am hopeful that AVB support in Pipewire will eventuate fully in the future. 

On macOS I observe glitches - restarts and buffer issue that I've not managed to eliminate. 

On Windows, WDM isn't an option with CamillaDSP as the 32 channels I need are presented as 4x8ch.

I've tried another convolver on Windows with ASIO and I observe very solid performance. No evidence of glitches seen on macOS. 

So, I do have a workable solution, of course would prefer to use CamillaDSP.

I'll watch this issue and the repo generally if the ASIO CPAL backend becomes an option in the future!

**Comment by HEnquist** [comment]
> On macOS I observe glitches - restarts and buffer issue that I've not managed to eliminate.

This doesn't usually happen on macOS, it's normally very reliable there. Is it easy for you to switch to macOS to run some tests? I would be interested in what the camilladsp logs say when there are problems.

**Comment by HEnquist** [comment]
I updated the old experimental ASIO branch to try. Things are somewhat improved since my last try, but it's not bug free and there are some quite severe limitations.

This is in the `next30-asio` branch here: https://github.com/HEnquist/camilladsp/tree/next30-asio

The ASIO api does not like the way that CamillaDSP accesses devices from different threads, meaning that it's only possible to use ASIO for capture or playback, not both at the same time.
See the readme section here: https://github.com/HEnquist/camilladsp/tree/next30-asio?tab=readme-ov-file#experimental-asio-support

In short I don't think that this works very well currently and I'm hesitant to include it. But maybe it's still useful for someone.

**Comment by siraaris** [comment]
Thank you, and sorry, I missed this update. I've got a dev environment on Windows now, so will give this a try.

**Comment by siraaris** [comment]
> > On macOS I observe glitches - restarts and buffer issue that I've not managed to eliminate.
> 
> This doesn't usually happen on macOS, it's normally very reliable there. Is it easy for you to switch to macOS to run some tests? I would be interested in what the camilladsp logs say when there are problems.

I have to setup my laptop a bit, but shouldn't be difficult. Will send logs over.

**Comment by siraaris** [comment]
Compiled next30-asio branch, including the dependency requirements. I modified the  .YML file from the version I was using on macOS, and get the following - seems to be stuck on a particular filter (KS-T).

[camilladsp-asio-log.txt](https://github.com/user-attachments/files/15846614/camilladsp-asio-log.txt)

**Comment by siraaris** [comment]
Config file attached.
[cexe.yml.txt](https://github.com/user-attachments/files/15846621/cexe.yml.txt)

**Comment by siraaris** [comment]
Ok, this is very weird. After navigating to the directory where the filter files are (File Explorer), and started CamillaDSP again, it works (or at least progresses) further now. I really don't understand why this would be happen, but happy that things are progressing, but still weird.

**Comment by siraaris** [comment]
Another error:

fatal runtime error: Rust cannot catch foreign exceptions

Something is definitely not right.

Attaching another trace when it starts, but something's not right with the rate (I've set it to 96k but reporting 16k, 48k etc).

[Uploading camilladsp-asio-log-sort-of-working.txt…]()

**Comment by siraaris** [comment]
Thanks for creating the branch! That was fun. Tantalisingly close, but not quite :-)

**Comment by siraaris** [comment]
I should have read the readme sorry.

**Comment by HEnquist** [comment]
> I should have read the readme sorry, clearly I need to use a virtual device not capture and playback on the same.

Yeah this is a pretty significant limitation, and it's really not an easy one to get around. Capture and playback needs to be done in the same thread, not separate threads like camilladsp does it now (which works fine for all the other audio APIs).

**Comment by siraaris** [comment]
I used a Windows loopback program (https://www.nerds.de/en/loopbeaudio.html) to handle 2ch input via Roon to bring into Camilla as a 2ch (Wasapi) capture, and using Asio as the (32ch) playback.

Now the app runs, and processes but I notice:

- Lots of dropped frames from Wasapi
- Variable rates from Wasapi

Both of these are weird as the rate is fixed in Windows sound control panel and Roon resamples to the chosen rate (96khz).

Hope the logs provide some indication of what's going on or potential hint at ways to fix.

[camilla-asio-wasapi.log.txt](https://github.com/user-attachments/files/15849470/camilla-asio-wasapi.log.txt)

**Comment by HEnquist** [comment]
This is a problem:
```
Processing load: 613.6589%
```
This basically means that the CPU is overloaded by more than 6x. I see you have a lot of FIR filters, but not so many that I would expect that huge load. The reason is probably that you are using an unusual chunksize of 19200. Try with a power of two instead, for example 16384. That should make the convolution run much faster.

**Comment by siraaris** [comment]
I've tried with lowering the chunk size to 16k. I'm not sure that the reported Processing load is overly meaningful as when watching Performance Monitor it was well under 10% when Camilla was reporting 613% - weird.

I also tried a different WDM based Loopback (VAC, https://vac.muzychenko.net/en/) which seems to be better behaved compared to the nerds.de one. If the chunksize is 16k, I don't get wasapi errors.

But regardless - I still get timeouts/restarts from the ASIO playback. 

Interestingly another (commercial) Windows convolver provides a virtual ASIO driver as an option to get audio into the convolver, and then the capture is taken from the virtual ASIO driver mirror of whatever's played into it.

That might be an interesting way (in the future perhaps) to consider - would then enable multithreaded capture and playback?

**Comment by HEnquist** [comment]
> I've tried with setting the chunk size to 16k. I'm not sure that the reported Processing load is overly meaningful as when watching Performance Monitor it was well under 10% when Camilla was reporting 613% - weird.

100% in Performance Monitor means that all all cores are fully utilised. If a single core on for example an 8-core machine is maxed out you will see 12.5 %. CamillaDSP simply measures how long it takes to process a chunk and compares with the duration of a chunk. Over 100% means it cannot keep up since processing runs slower than real time.
What CPU did you try this on?

> 
> I also tried a different WDM based Loopback (VAC, https://vac.muzychenko.net/en/) which seems to be better behaved compared to the nerds.de one. If the chunksize is 16k, I don't get wasapi errors.

Good tip, thanks!

> 
> But regardless - I still get timeouts/restarts from the ASIO playback.
> 
> Interestingly another (commercial) Windows convolver provides a virtual ASIO driver as an option to get audio into the convolver, and then the capture is taken from the virtual ASIO driver mirror of whatever's played into it.
> 
> That might be a way (in the future perhaps) to consider - would then enable multithreaded capture and playback?

That would be possible yes, in theory at least. There are also issues with the ASIO SDK license. It's very restrictive and doesn't fit well in open source projects. This means my motivation for working with it is quite low unfortunately.

**Comment by siraaris** [comment]
Ok, I'll tinker with Wasapi capture and Asio playback a bit more and see what I find.

**Comment by siraaris** [comment]
Not much to report. I downloaded and tried out VB-Cable, no real difference to VAC though.

If logs are of interest please let me know and I'll run a trace and attach.

**Comment by HEnquist** [comment]
Does it run reliably with a lightweight config? For example if you keep only the mixer(s) in the pipeline but remove all filters.

**Comment by siraaris** [comment]
Yes, with no filters it runs stable. Attaching three configs and log scenarios:

01 - Config with no filters, just mixers - stable, clearly single is presented to DAC (ie TotalMix reports signal on all 32 ch in use).

02 - Config with Gain and Delay added to 1st-stage mixer (2 in, 18 out). Looks like filtering loops quite a few times, adding each to the whole set? Could just be logging. Rate is observed as set (96khz). However no signal is presented to DAC on any channels.

03 - Config with only FIR filters enabled on the 2nd-stage mixer (18 in, 32 out). Same looping of filters, plus a bunch of logs "Dropping captured chunk XXX with len 7680". Samplerate is not correctly calculated.

[01_YML_3sb-plain-mixer.txt](https://github.com/user-attachments/files/15967021/01_YML_3sb-plain-mixer.txt)
[01_LOG_camilla-asio-plain-mixer.txt](https://github.com/user-attachments/files/15967015/01_LOG_camilla-asio-plain-mixer.txt)

[02_YML_3sb-gain+delay.txt](https://github.com/user-attachments/files/15967019/02_YML_3sb-gain%2Bdelay.txt)
[02_LOG_camilla-asio-gain+delay.txt](https://github.com/user-attachments/files/15967016/02_LOG_camilla-asio-gain%2Bdelay.txt)

[03_YML_3sb-fir.txt](https://github.com/user-attachments/files/15967020/03_YML_3sb-fir.txt)
[03_LOG_camilla-asio-fir.txt](https://github.com/user-attachments/files/15967018/03_LOG_camilla-asio-fir.txt)

**Comment by HEnquist** [comment]
> using the aggregated pipeline Filter groupings

Ah that explains it! There is a bug, the old "channel" property on filter steps has been removed. It's supposed to give an error if the config has it, but that is missing. The result then is that it ignores the "channel" property. And when "channels" (with the s) is missing that means apply to all channels. So all filters got applied to all channels, no wonder it got a bit heavy..

I'll fix the bug asap.

**Comment by siraaris** [comment]
I've been running now for a few hours. Any plans to include ASIO support into next30?

**Comment by HEnquist** [comment]
I'm not planning to include it in the next release, there are too many problems with it. But I'll update the asio branch to have all the new things from next30.

**Comment by siraaris** [comment]
Thank you! I can probably track if you decide to leave it u til next40. I'll submit any issue I find :)

**Comment by siraaris** [comment]
By the way, when I enabled (just for kicks) the FFTW feature, camilladsp.exe just exits, with no errors even with -vv
Seems weird, although I am not going to use the feature, RustFFT is fine.

Also, how do I stop camilladsp.exe exiting in a PowerShell session when I disconnect from a Remote Desktop session?

**Comment by HEnquist** [comment]
Just tried the FFTW feature on an arm windows system, and there it doesn't build at all. The FFTW rust binding depends on some very outdated libraries and it doesn't look like it's likely to be fixed anytime soon. I'll just remove it.

I don't know about remote desktop, but could it be related to this? There are some suggestions on the last page.
https://answers.microsoft.com/en-us/windows/forum/windows8_1-networking/remote-desktop-connection-does-not-leave-programs/f29719e9-64d2-4614-b66c-2363511a9293?page=1

**Comment by siraaris** [comment]
It's weird. If I have another process eg more file.yml, and disconnect RDP and reconnect that process remains. It's only camilladsp.exe that dies. Almost like RDP disconnect sends a signal to the process. 

Actually I maybe think it's to do with the audio Loopback device being removed with RDP disconnect.

**Comment by siraaris** [comment]
Logs:

2024-06-27 16:39:57.228136 TRACE [src\wasapidevice.rs:493] capturing
2024-06-27 16:39:57.228147 TRACE [src\wasapidevice.rs:1068] got chunk, length 7680 bytes
2024-06-27 16:39:57.228179 TRACE [src\wasapidevice.rs:1065] capture device needs more samples to make chunk, reading from channel
2024-06-27 16:39:57.239172 TRACE [src\wasapidevice.rs:527] Available frames from capture dev: 960
2024-06-27 16:39:57.239217 TRACE [C:\Users\Aris Theocharides\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasapi-0.15.0\src\api.rs:1165] read 960 frames
2024-06-27 16:39:57.239232 TRACE [src\wasapidevice.rs:493] capturing
2024-06-27 16:39:57.239248 TRACE [src\wasapidevice.rs:1068] got chunk, length 7680 bytes
2024-06-27 16:39:57.239303 TRACE [src\wasapidevice.rs:1065] capture device needs more samples to make chunk, reading from channel
2024-06-27 16:39:57.249501 TRACE [src\wasapidevice.rs:527] Available frames from capture dev: 960
2024-06-27 16:39:57.249543 TRACE [C:\Users\Aris Theocharides\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasapi-0.15.0\src\api.rs:1165] read 960 frames
2024-06-27 16:39:57.249556 TRACE [src\wasapidevice.rs:493] capturing
2024-06-27 16:39:57.249573 TRACE [src\wasapidevice.rs:1068] got chunk, length 7680 bytes
2024-06-27 16:39:57.249612 TRACE [src\wasapidevice.rs:1065] capture device needs more samples to make chunk, reading from channel
2024-06-27 16:39:57.260396 TRACE [src\wasapidevice.rs:527] Available frames from capture dev: 960
2024-06-27 16:39:57.260431 TRACE [C:\Users\Aris Theocharides\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasapi-0.15.0\src\api.rs:1165] read 960 frames
2024-06-27 16:39:57.260439 DEBUG [src\wasapidevice.rs:553] Captured a buffer marked as silent
2024-06-27 16:39:57.260450 TRACE [src\wasapidevice.rs:493] capturing
2024-06-27 16:39:57.260492 TRACE [src\wasapidevice.rs:1068] got chunk, length 7680 bytes
2024-06-27 16:39:57.260523 TRACE [src\wasapidevice.rs:1065] capture device needs more samples to make chunk, reading from channel
2024-06-27 16:39:57.266040 TRACE [src\wasapidevice.rs:527] Available frames from capture dev: 960
2024-06-27 16:39:57.266071 TRACE [C:\Users\Aris Theocharides\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasapi-0.15.0\src\api.rs:1165] read 960 frames
2024-06-27 16:39:57.266078 DEBUG [src\wasapidevice.rs:553] Captured a buffer marked as silent
2024-06-27 16:39:57.266103 TRACE [src\wasapidevice.rs:493] capturing
2024-06-27 16:39:57.266114 TRACE [src\wasapidevice.rs:1068] got chunk, length 7680 bytes
2024-06-27 16:39:57.266149 TRACE [src\wasapidevice.rs:1065] capture device needs more samples to make chunk, reading from channel
2024-06-27 16:39:57.287222 TRACE [C:\Users\Aris Theocharides\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasapi-0.15.0\src\events.rs:157] state change to: Inactive
2024-06-27 16:39:57.297598 TRACE [C:\Users\Aris Theocharides\.cargo\registry\src\index.crates.io-6f17d22bba15001f\wasapi-0.15.0\src\events.rs:174] Disconnected
2024-06-27 16:39:57.297629 DEBUG [src\wasapidevice.rs:460] Capture disconnected, reason: SessionDisconnected
2024-06-27 16:39:57.297853 ERROR [src\wasapidevice.rs:1093] Channel is closed
2024-06-27 16:39:57.297884 DEBUG [src\wasapidevice.rs:878] Send CaptureError
2024-06-27 16:39:57.297887 TRACE [src\processing.rs:34] AudioMessage::EndOfStream received
2024-06-27 16:39:57.297946 ERROR [src/bin.rs:316] Capture error: receiving on an empty and disconnected channel
2024-06-27 16:39:57.302042 DEBUG [src/bin.rs:321] Wait for playback thread to exit..
2024-06-27 16:39:57.576687 TRACE [src/bin.rs:330] All threads stopped, returning
2024-06-27 16:39:57.576791 DEBUG [src/bin.rs:1136] Processing ended with status Ok(Restart)
aris@CONTROLLER:~$

**Comment by siraaris** [comment]
Ok, known issue.

https://vac.muzychenko.net/en/manual/compat.htm#:~:text=To%20access%20Virtual%20Cable%20devices,the%20Remote%20Desktop%20connection%20dialog.

**Comment by siraaris** [comment]
So, for anyone using RDP - swap or add VNC Server, and that will hold the WASAPI devices present so when you disconnect the loopback audio will remain and continue to work.

A bit of a kludge, but it works well.

**Comment by siraaris** [comment]
Looks like camillagui needs slight modification to support Asio devices. A quick attempt looks like adding support for Asio is relatively straightforward, but there are still some issues, so I think that the next30 branch of camillagui is still in progress...

**Comment by HEnquist** [comment]
Yes the gui and the pycamilladsp-plot library need some changes to add the Asio option. There should not be any major issues with the next30 branch of the gui, did you try it with next30 of everything? I mean pycamilladsp, pycamilladsp-plot, gui and gui backend.

**Comment by siraaris** [comment]
Yes, I tried it with next30 on everything. I tried to fix the missing support for Asio devices, but I'm sure that I didn't do it correctly. I actually don't use the UI, was just looking at it as a matter of curiosity where it was at (as some audio DSP friends  are interested - but I think they will wait happily until 3.0 is released proper).

**Comment by MarcoRavich** [comment]
Hi everyone,
since I'm really interested in (native) ASIO support, I strongly encourage @HEnquist to establish some kind of collaboration with other open source projects' authors in this field.

I've collected some resources about for the [HyMPS project](https://forart.it/HyMPS) under [AUDIO section](https://github.com/FORARTfe/HyMPS#-) \ [Tools page](https://github.com/FORARTfe/HyMPS/blob/main/Audio/Tools.md#--).

Hope that helps/inspires.

(note: CamillaDSP is listed under [System wide subsection](https://github.com/FORARTfe/HyMPS/blob/main/Audio/Tools.md#system-wide-), should I change the placement ?)

**Comment by HEnquist** [comment]
Not much has changed since I wrote https://github.com/HEnquist/camilladsp/issues/346#issuecomment-2131146769

Available time is always limited and I'm not very keen to work with asio with it's old clunky api and annoyingly restrictive license.

**Comment by siraaris** [comment]
There some weird behaviour on Asio (Playback, to an RME device), and Wasapi (Capture, via VAC).

Pretty consistently after 7 minutes runtime, I get "Samples were dropped" - by the Wasapi device:

2025-03-12 02:40:11.743118 WARN [src\wasapidevice.rs:1129] Samples were dropped, missing 1 buffers
2025-03-12 02:40:14.473761 WARN [src\wasapidevice.rs:1129] Samples were dropped, missing 1 buffers
2025-03-12 02:40:17.204432 WARN [src\wasapidevice.rs:1129] Samples were dropped, missing 1 buffers
2025-03-12 02:40:18.783230 WARN [src\wasapidevice.rs:1129] Samples were dropped, missing 1 buffers
2025-03-12 02:40:20.831061 WARN [src\wasapidevice.rs:1129] Samples were dropped, missing 1 buffers
2025-03-12 02:40:23.604513 WARN [src\wasapidevice.rs:1129] Samples were dropped, missing 1 buffers
2025-03-12 02:40:25.780641 WARN [src\wasapidevice.rs:1129] Samples were dropped, missing 1 buffers

The error (and audible glitches) continue on, and the program needs to be stopped/restarted.

2025-03-12 02:40:27.043522 INFO [src\bin.rs:1021] Shutting down


Not sure this is a Wasapi issue or Asio?

Aris

**Comment by HEnquist** [comment]
Do you have an async resampler enabled? I not, it might just be a rare mismatch between capture and playback.

**Comment by siraaris** [comment]
I do have an async resample enabled yes. A trace log is available [here](https://www.dropbox.com/scl/fi/f9kcr6hmdp0xz0psazh2l/camilladsp.log?rlkey=spaqb7wlmkj0f216yn0q9mwp0&st=0ktm0jvk&dl=0).

**Comment by HEnquist** [comment]
Ok looks like there is something wrong with the buffer level measurement. It gives incorrect very small values, and that makes the rate adjust speed up, and after a while the channels are full so it starts dropping data.
I'll fix it next time I do something with asio. It's low on the priority list though so no promises for when that will be.

**Comment by siraaris** [comment]
Ok, that makes sense. If there's anything I can do, things to check or test, just let me know. I checked in on ASIO as there are a few folks here in Australia that I'd like to get set up with Camilla, but there's no real hurry.

On the plus side - Asio (playback) seems pretty stable.

**Comment by HEnquist** [comment]
The buffer level measurement for s on tje playback side, so must be in Asio. It might also affect Jack, that's something I will need to look into.

**Comment by HEnquist** [comment]
I have seen this! It's available with a GPLv3 license, far better that it was. CamillaDSP until V3 is also GPLv3, but for V4 it changes to a more permissive dual licence of GPLv3 and MPL 2.0. This makes things a little tricky, as soon as an ASIO backend is included the entire project must drop the MPL license.
This can probably be solved by structuring the code in some smart way so that it can still be distributed as dual licensed without ASIO.

So the current state is that the license situation has improved (but is still non-trivial), and the ASIO api is still difficult to work with.

**Comment by HEnquist** [comment]
There is a new asio backend in branch asio_testing! It's very new and mostly untested, but this has much better chances of working well than going via CPAL.
It supports full duplex mode, so the same device can be used for playback and capture. Using different ASIO devices for playback and capture is not supported, but combining with a different backend is of course possible.

**Comment by HEnquist** [comment]
There is a binary with ASIO support published here (see the bottom of the page):
https://github.com/HEnquist/camilladsp/actions/runs/22594945359

**Comment by Frunsianer** [comment]
Hello Henrik, i will try the ASIO Branch on Windows 10 with an RME HDSPe AES PCIEX1 Card 16x16 

For now i got Camilla running with Wasapi Fooobar2000 to CamillaDSP to HDMI Out 2 Chanels.
Next step is to build ab an stand alone PC with used only for Camilla and that  RME CARD with ASIO in and Out for Surround pass thrue with full aktive 7.2 Setup.

You can find that Project here (work in Progress): 

[https://www.diyaudio.com/community/threads/pc-based-dsp-with-rme-hdspe-aes-and-camilladsp-project-for-full-surround-digital-aktive-setup.440807/](url)

**Comment by Frunsianer** [comment]
And thank you for youre great Work! 
An alterntaive on Windows would be EKio but it had now FIR Filter Suppord and it isn't free for more than 4 CHanels Out.

But even youre Software would cost money i would use it!!!
The Target is to Run Camilla on an RPI5 with that Card and an PCIe 1X Addonboard for an smal Formfactor DSP 16 in and 16 Out! with Masterclock in and Out possebility!

**Comment by Frunsianer** [comment]
Damm i read this right now:
"The ASIO API has a number of limitations. In CamillaDSP the capture and playback devices are initialized and accessed from two separate threads. This is not compatible with the ASIO API, meaning that it's not possible to use ASIO for both playback and capture simultaneosly. To use an ASIO playback device, it is therefore necessary to use a Wasapi capture device (or vice versa)."

Is that still not possible to use the Same Device for in and Out thrue ASIO ??? 

Huston i got an Problem. :(

Need to check if the RME Driver has Waspi support as well.

**Comment by Frunsianer** [comment]
Found this:
"The RME HDSPe AES supports WASAPI on Windows through its WDM driver layer. However, the official RME drivers are heavily optimized for ASIO. While WASAPI Exclusive mode will bypass the Windows mixer, the driver does not natively allow applications to automatically change sample rates"

**Comment by HEnquist** [comment]
It looks like you are looking at the very outdated old ASIO branch. ASIO is included in the latest version. See https://github.com/HEnquist/camilladsp#asio
There is also a pre-built binary with ASIO support enabled, see https://github.com/HEnquist/camilladsp/releases/tag/v4.1.3

**Comment by Frunsianer** [comment]
Thank You, and i also used that Branch for Windows without that  Asio in the Fillename ;) LOL

No Wonder Asio mad Problem's!

**Comment by Frunsianer** [comment]
No i had the right Version already the File hase aexctly the same Sice  7.384KB
And i had already an Folder with the ASIO in his Name.

But i have some problems still with smal little droppouts thrue Camilla barely Hearable silend pops and kracks!
It Sounds not Clean!

Foobar  (WASAPI: CABLE Input (VB-Audio Virtual Cable) 
CamilaDSP IN Device (WASAPI: CABLE Output (VB-Audio Virtual Cable)
CamillaDSP OUT Device (Realtek Digital Output (Realtek USB Audio))

Mixer is only 2 Chanel in and Out 
one Filter in Between for Baffelstep compensation (-6db)


If i use the Realtek Digital Output (Realtek USB Audio) Device on Foobar directly much better Sound and no crackling in between or pops or silend dropouts!

I didn't had this on Linux with my MOTU UL MK5 on USB 

Not sure what could be the reason my be that VB-AUDIO Device!

**Comment by Frunsianer** [comment]
Let see how this would run without this VB-AUDIO device.
Next day i get all my PC Parts for my Standalone DSP Device and i will test it there directly with the RME Card IN AND OUT nothing else involved thrue ASIO, on AES Outs the same DAC's

**Comment by Frunsianer** [comment]
Hello Henrik, i got my RME Card running with Windows and CamillaDSP (ASIO), but i have an big problem, the Mixer don't do anything!!!

If i delete the Mixer the Musik still playing 2 Channels in 2 Channels Out, so i cant setup Crossover this way.

But if i add Filters they work! So this way i have only 16 IN starit to 16 Out with Filters but no Mixing Possible !! :( 

I need 8 IN  mixed to 16 OUT ;)

I tried the same Setup wit EKIO DSP and his Mixer works perfectly on ASIO so i guess there is an Bug or so, on RME Mixer i use the Same Settings for Both Setups.

I Use the Audio Workstation Mode on RME Card , no Input Mix, Playback Routed to Outs strait.

**Comment by Frunsianer** [comment]
Or my Setup File has an Problem on  CamillaDSP? 

But i didn't see an Error there.
Is still playing only Stereo in and Out 2 Chanels used at all.

mixers:
  Test:
    channels:
      in: 16
      out: 16
    description: null
    labels: null
    mapping:
    - dest: 0
      mute: false
      sources:
      - channel: 0
        gain: 0
        inverted: false
        mute: false
        scale: dB
    - dest: 1
      mute: false
      sources:
      - channel: 0
        gain: 0
        inverted: false
        mute: false
        scale: dB
    - dest: 2
      mute: false
      sources:
      - channel: 1
        gain: 0
        inverted: false
        mute: false
        scale: dB
    - dest: 3
      mute: false
      sources:
      - channel: 1
        gain: 0
        inverted: false
        mute: false
        scale: dB

**Comment by HEnquist** [comment]
This issue is specifically about ASIO support, and it is already closed.
GitHub Issues are useful for bugs, regressions, and discussion of new features. I do not use them for general support. For configuration or setup questions, please use the diyaudio thread. There are many active users there who can help, and replies are much faster. Here it is basically just me, so you may end up waiting a long time for an answer.

Please also put configuration snippets inside code blocks so they are readable.

**Comment by Frunsianer** [comment]
Ok i will do, thank You
